cassiopeia/solver/UPGMASolver.py

```diff
@@ -0,0 +1,213 @@
+"""
+This file stores a subclass of DistanceSolver, UPGMA. The inference procedure is
+the Neighbor-Joining algorithm proposed by Sokal and Michener (1958) that 
+iteratively joins together samples with the minimum dissimilarity.
+"""
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+import abc
+import networkx as nx
+import numba
+import numpy as np
+import pandas as pd
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.solver import DistanceSolver
+
+
+class UPGMASolver(DistanceSolver.DistanceSolver):
+    """UPGMA class for Cassiopeia.
+
+    Implements the UPGMA algorithm described as a derived class of 
+    DistanceSolver. This class inherits the generic `solve` method, but 
+    implements its own procedure for finding cherries by minimizing the 
+    dissimilarity between samples. After joining nodes, the dissimilarities
+    are updated by averaging the distances of elements in the new cluster 
+    with each existing node. Produces a rooted tree that is assumed to be
+    ultrametric.
+
+
+    Args:
+        dissimilarity_function: A function by which to compute the dissimilarity
+            map. Optional if a dissimilarity map is already provided.
+        prior_transformation: Function to use when transforming priors into
+            weights. Supports the following transformations:
+                "negative_log": Transforms each probability by the negative
+                    log (default)
+                "inverse": Transforms each probability p by taking 1/p
+                "square_root_inverse": Transforms each probability by the
+                    the square root of 1/p
+
+    """
+
+    def __init__(
+        self,
+        dissimilarity_function: Optional[
+            Callable[
+                [np.array, np.array, int, Dict[int, Dict[int, float]]], float
+            ]
+        ] = None,
+        prior_transformation: str = "negative_log",
+    ):
+
+        super().__init__(
+            dissimilarity_function=dissimilarity_function,
+            add_root=True,
+            prior_transformation=prior_transformation,
+        )
+
+
+    def root_tree(self, tree: nx.Graph, root_sample: str, remaining_samples: List[str]):
```

Comment:
Should we make `root_tree` a private method? That way we don't need to really explain to users why some arguments are used or not used.

Generally, this would be cleaned up with Sebastian's suggestion of taking a compositional approach for proposing a root. Let's make a note of that!

Reply:
So I think that inheritance with private methods is throwing me some errors, so I left it as a public method for now